### ISSUE TITLE
[QUESTION] Is there a common pattern for negative mining?

### ISSUE THREAD

**Issue by Iambestfeed** [UKPLab/sentence-transformers#3279]
### **Context**  
I have a **query-passage dataset** and applied several **filtering techniques** to evaluate their impact. Based on my findings, I formulated a **data filtering strategy**, which reduced the dataset from **300k** to **70k samples**.  

### **Key Insights from Research**  
📌 **From Blog [1]:** Selecting **negatives from rank 30+** improves performance.  
📌 **From Issue [2]:** Using **negatives worse than pre-trained data** may **harm performance**.  
📌 **From My Experiment:**  
   - Choosing **negatives from top 30+** prevents excessive semantic similarity.  
   - I applied **near-deduplication** to enhance **batch negative sampling** effectiveness.  

### **Questions**  
1️⃣ **What is the most effective strategy for building hard negatives, based on your experience?**  
2️⃣ Given that **rerank models in my language are weak**, should I mine negatives using **BM25** or leverage a **biencoder (e.g., BGE-M3)?**  
3️⃣ If I limit negatives to **top 10 or fewer (instead of top 30)**, will this make the dataset **too challenging** for the model?  

**References:**  
[1] [[Salesforce Blog on Embeddings](https://www.salesforce.com/blog/sfr-embedding/)](https://www.salesforce.com/blog/sfr-embedding/)  
[2] [[UKPLab Issue on Sentence Transformers](https://github.com/UKPLab/sentence-transformers/issues/2358)](https://github.com/UKPLab/sentence-transformers/issues/2358)

**Comment by Iambestfeed** [comment]
@tomaarsen Sorry for tagging you but I would love to hear about your experiences with this problem.

**Comment by tomaarsen** [comment]
Hello!

Big apology for the delay, the v4.0 release took my attention away from the issues for a minute.

To answer your questions:
1. What is the most effective strategy for building hard negatives, based on your experience?

This is a difficult question, with a lot of research behind it already, e.g. the [NV-Retriever paper](https://arxiv.org/abs/2407.15831). In my personal experience, harder is almost always better, but I never use hard negatives that are deemed more similar than the positive. In the `mine_hard_negatives` function, this means `margin=0`.

According to this paper, there are some other valuable learnings:
a. Stronger embedding models used in hard negatives mining results in stronger models.
b. Ensembling hard negatives from different models didn't help much over using a strong base model
c. The strongest performing seems to be to take the top $k$ hard negatives that are at most 95% as similar as the positive sample. (outperforming e.g. absolute margins to the positive, absolute similarity scores, skipping the top K samples, etc.)

I see now that this `relative_margin` isn't implemented in `mine_hard_negatives`. I'll add that now.

2. Given that rerank models in my language are weak, should I mine negatives using BM25 or leverage a biencoder (e.g., BGE-M3)?

Hard negatives mining is often done using bi-encoders, so I reckon a multilingual (or monolingual in your language) biencoder should work very nicely. It'll likely outperform BM25 unless the biencoders are also poor on your language.

3. If I limit negatives to top 10 or fewer (instead of top 30), will this make the dataset too challenging for the model?

I don't think so, as long as you also ensure that e.g. the negatives aren't stronger than your positive, for example by setting the margin to 0 or 0.05.

- Tom Aarsen

**Comment by bobox2997** [comment]
@tomaarsen out of curiosity, have you tried to use this margin with some 'modern' model like artic embedding or bge? seems that the similarity of many sentences is really 'clustered', and ie while running it on natural questions dataset using artic embedd m and, relative margin of 0.05, and num negative of 5, it fail to find the enough negatives for a relevant portion of the dataset (the average similarity for positive and negative is somehow similar).

I suppose this is related to the fact that those models are trained with really high scaling factor in cross entropy loss (or low temp if we see it as it is implemented in losses like GISTEmbed loss).

I had to change the model to all mpnet v2 otherwise I would have to reduce / remove the margin and I'm not confortable with that.

seeing that with hard_negative_mining, I'm wondering if that's happening also inside my training with GISTEmbed loss, as I use artic embedd s as 'guide' model with 0.05 relative margin. 
as a really, fast, approximate check I tried changing it with all-MiniLM-L6-v2 and the training loss increased (still I runned just a few steps as I don't have time right now) 

when I have some time I'll edit the loss function to print the number of removed negative per batch, unfortunately I will be too busy for the next ~10 days. 

i just ended up seeing that issue here and I tought that this observation may be useful (I myself didn't noticed that until I tried to use those models to mine hard negatives)

**Comment by tomaarsen** [comment]
I haven't used it with Arctic or BGE I think, no. I'm surprised that you can't find many negatives though, I feel like relative margin of 0.05 should be fine for those models too. Perhaps the [range_max](https://sbert.net/docs/package_reference/util.html#sentence_transformers.util.mine_hard_negatives) is too low for your dataset? I.e. it considers 50 negatives, but all 50 fall within the 95% of the `positive`. Increasing `range_max` can help. 

- Tom Aarsen

**Comment by bobox2997** [comment]
I'll try, I used the default 50. 
runned it with relative margin 0.033 and absolute margin of 0.02, max negative 50 and it cannot select 3 negative for 50k examples using artic embedd l. same settings, all-mpnet v2 fount 99k.
Probably you are right, increasing max negatives would find more, but this mean that the margin is highly relative and using a 'better' model wouldn't necessarily found better negatives

**Comment by bobox2997** [comment]
@tomaarsen 

i've made some tests...
here the stats for artic-embedd-m:

biencoder_name = "Snowflake/snowflake-arctic-embed-m"
dataset = load_dataset("sentence-transformers/natural-questions", split="train")

max_score=None
num_negatives=3
relative_margin=0.05
absolute_margin=None

range_max=50

Metric       Positive       Negative     Difference
Count         100,231         65,331               
Mean           0.6296         0.6977         0.0496
Median         0.6343         0.6963         0.0406
Std            0.0938         0.0343         0.0228
Min            0.1827         0.5395         0.0300
25%            0.5674         0.6746         0.0376
50%            0.6343         0.6963         0.0406
75%            0.6942         0.7199         0.0503
Max            0.9462         0.8823         0.2661
Skipped 4,304,618 potential negatives (84.21%) due to the relative_margin of 0.05.
Could not find enough negatives for 235362 samples (78.27%).


range_max=100

Metric       Positive       Negative     Difference
Count         100,231         87,732               
Mean           0.6296         0.6874         0.0460
Median         0.6343         0.6849         0.0385
Std            0.0938         0.0364         0.0206
Min            0.1827         0.5376         0.0286
25%            0.5674         0.6623         0.0361
50%            0.6343         0.6849         0.0385
75%            0.6942         0.7108         0.0453
Max            0.9462         0.8823         0.2661
Skipped 7,997,990 potential negatives (79.01%) due to the relative_margin of 0.05.
Could not find enough negatives for 212961 samples (70.82%). 

range_max = 250

Metric       Positive       Negative     Difference
Count         100,231        125,046               
Mean           0.6296         0.6719         0.0424
Median         0.6343         0.6681         0.0365
Std            0.0938         0.0404         0.0181
Min            0.1827         0.5330         0.0282
25%            0.5674         0.6429         0.0344
50%            0.6343         0.6681         0.0365
75%            0.6942         0.6986         0.0411
Max            0.9462         0.8823         0.2661
Skipped 17,566,878 potential negatives (69.83%) due to the relative_margin of 0.05.
Could not find enough negatives for 175647 samples (58.41%). 

range_max = 1000

Metric       Positive       Negative     Difference
Count         100,231        193,164               
Mean           0.6296         0.6455         0.0386
Median         0.6343         0.6417         0.0343
Std            0.0938         0.0504         0.0155
Min            0.1827         0.4973         0.0263
25%            0.5674         0.6063         0.0321
50%            0.6343         0.6417         0.0343
75%            0.6942         0.6809         0.0379
Max            0.9462         0.8823         0.2661
Skipped 51,156,260 potential negatives (50.99%) due to the relative_margin of 0.05.
Could not find enough negatives for 107529 samples (35.76%). 


I'll make some tests using other models, but as pointed out in e5 repos:

Why does the cosine similarity scores distribute around 0.7 to 1.0?
This is a known and expected behavior as we use a low temperature 0.01 for InfoNCE contrastive loss.
For text embedding tasks like text retrieval or semantic similarity, what matters is the relative order of the scores instead of the absolute values, so this should not be an issue.

so probably the margin is really relative, depending on models
